optuna/terminator/improvement/evaluator.py

```diff
@@ -105,3 +103,101 @@ def _validate_input(
+    """Convert a list of FrozenTrial objects to tensors inputs and bounds.
+
+    This function assumes the following condition for input trials:
+    - any categorical param is converted to a float or int one;
+    - log is unscaled for any float/int distribution;
+    - the state is COMPLETE for any trial;
+    - direction is MINIMIZE for any trial.
+    """
+    search_space = IntersectionSearchSpace().calculate(trials)
+    sorted_params = sorted(search_space.keys())
+
+    x = []
+    for trial in trials:
+        assert trial.state == TrialState.COMPLETE
+        x_row = []
+        for param in sorted_params:
+            distribution = search_space[param]
+
+            assert not _distribution_is_log(distribution)
+            assert not isinstance(distribution, CategoricalDistribution)
+
+            param_value = float(trial.params[param])
+            x_row.append(param_value)
+
+        x.append(x_row)
+
+    min_bounds = []
+    max_bounds = []
+    for param, distribution in search_space.items():
+        assert isinstance(distribution, (FloatDistribution, IntDistribution))
+        min_bounds.append(distribution.low)
+        max_bounds.append(distribution.high)
+    bounds = [min_bounds, max_bounds]
+
+    y = torch.tensor([trial.value for trial in trials], dtype=torch.float64)
+
+    return torch.tensor(x, dtype=torch.float64), torch.tensor(bounds, dtype=torch.float64), y
+
+
+def _get_beta(n_params: int, n_trials: int, delta: float = 0.1) -> float:
+    beta = 2 * np.log(n_params * n_trials**2 * np.pi**2 / 6 / delta)
+
+    # The following div is according to the original paper: "We then further scale it down
+    # by a factor of 5 as defined in the experiments in Srinivas et al. (2010)"
+    beta /= 5
+
+    return beta
+
+
+def _fit_gp(x: torch.Tensor, bounds: torch.Tensor, y: torch.Tensor) -> SingleTaskGP:
+    y = torch.unsqueeze(y, 1)
+    gp = SingleTaskGP(
+        x,
+        y,
+        input_transform=Normalize(d=x.shape[1], bounds=bounds),
+        outcome_transform=Standardize(m=1),
+    )
+    mll = gpytorch.mlls.ExactMarginalLogLikelihood(gp.likelihood, gp)
+    fit_gpytorch_model(mll)
+    return gp
+
+
+def _calculate_min_lcb(
+    gp: SingleTaskGP, beta: float, x: torch.Tensor, bounds: torch.Tensor
+) -> float:
+    neg_lcb_func = UpperConfidenceBound(gp, beta=beta, maximize=False)
+
+    with gpytorch.settings.fast_pred_var():  # type: ignore[no-untyped-call]
+        x_opt, lcb = optimize_acqf(
+            neg_lcb_func,
+            bounds=bounds,
+            q=1,
+            num_restarts=10,
+            raw_samples=2048,
+            sequential=True,
+            options={"sample_around_best": True},
+        )
+        min_lcb = -lcb.item()
+        min_lcb_x = torch.min(-neg_lcb_func(x[:, None, :])).item()
+        min_lcb = min(min_lcb, min_lcb_x)
+
+    return min_lcb
+
+
+def _calculate_min_ucb(
+    gp: SingleTaskGP, beta: float, x: torch.Tensor, bounds: torch.Tensor
```

Comment:
`bounds` seems to be a redundant argument. Do you keep it for the consistency with `_calculate_min_lcb`?